### PR TITLE
#36869 Shotgun Pipeline Configuration Lookup

### PR DESCRIPTION
**Pull request by manneohrstrom**
This adds Shotgun runtime configuration lookup to the Toolkit boostrap manager.
This means that it's possible to implement dynamic overrides via the existing Pipeline Configuration concept in Shotgun, allowing for shotgun users to take control over the behaviour, configuration and deployment of plugin based payloads. Notable points:

- The system will look for the fields `entry_point` or `sg_entry_point` to determine the entry point. The latter field is to ensure backwards compatibility with older shotgun versions, where you can create this field by hand.
- The lookup will first check the traditional path fields, then look for the fields `descriptor` and `sg_descriptor` where it expects to find a descriptor uri string to resolve the config from
- The ToolkitManager.pipeline_configuration property has shifted its default value to be None rather than primary. If None, it will automatically determine the pipeline configuration to be used by looking at the users field in Shotgun. This makes it easy to support dev sandboxes for taap.

### CODE REVIEW

File: python/tank/bootstrap/manager.py [hunk 1]
```diff
@@ -64,15 +64,13 @@ def __repr__(self):
     def _get_pipeline_configuration(self):
         """
         The pipeline configuration that is being operated on.
-        By default, the primary pipeline config will be used.
+        By default, the manager will attempt to automatically determine
```
Comment (josh-t) [inline python/tank/bootstrap/manager.py]: Minor: These new docstring lines feels oddly placed. Seems like they'd be more at home in the class docstring.

File: python/tank/bootstrap/resolver.py [hunk 2]
```diff
@@ -131,36 +133,163 @@ def resolve_shotgun_configuration(
         self,
         pipeline_config_name,
         fallback_config_descriptor,
-        sg_connection
+        sg_connection,
+        current_login
```
Comment (josh-t) [inline python/tank/bootstrap/resolver.py]: Any worries about backward compatibility if anybody is using this? Could this be optional?

File: python/tank/bootstrap/resolver.py [hunk 3]
```diff
@@ -131,36 +133,163 @@ def resolve_shotgun_configuration(
...
+                    # user config
+                    if user_config:
+                        log.warning(
+                            "More than one user config detected. Will use the most "
+                            "recently updated one."
+                        )
+                    user_config = pc
+
+            # user pc takes precedence if available.
+            if user_config:
```
Comment (josh-t) [inline python/tank/bootstrap/resolver.py]: minor: i like this explicitness, but I now others who would suggest reducing this to one line.

File: python/tank/bootstrap/resolver.py [hunk 4]
```diff
@@ -131,36 +133,163 @@ def resolve_shotgun_configuration(
...
+                ],
+                fields,
+                order=[{"field_name": "updated_at", "direction": "asc"}]
+            )
+
+            log.debug(
+                "The following pipeline configurations were found: %s" % pprint.pformat(pipeline_configs)
+            )
+
+            for pc in pipeline_configs:
```
Comment (josh-t) [inline python/tank/bootstrap/resolver.py]: Could you set the order to descending and `break` after the first match?

File: python/tank/bootstrap/resolver.py (hunk 2 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/resolver.py]: the resolver is not part of the public API so should be ok

File: python/tank/bootstrap/resolver.py (hunk 3 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/resolver.py]: lol! who is this *others* person? Changed.

File: python/tank/bootstrap/resolver.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/resolver.py]: it's seemed nice to keep it consistent with the other case above though, where you have to check all of them....

File: python/tank/bootstrap/manager.py (hunk 1 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/manager.py]: i'll elaborate a bit!

**Pull request merged**